src/provider_diagnostics.jl

```diff
@@ -1,18 +1,117 @@
-const LintSeverity = Dict('E'=>1,'W'=>2,'I'=>3)
-
-function process_diagnostics(uri::String, server::LanguageServerInstance)
-    document = get_text(server.documents[uri])
-    L = lintfile(normpath(unescape(URI(uri).path))[2:end], String(document))
-    diags = map(L) do l
-        start_col = findfirst(i->i!=' ', get_line(uri, l.line, server))
-        Diagnostic(Range(Position(l.line-1, start_col-1), Position(l.line-1, typemax(Int)) ),
-                        LintSeverity[string(l.code)[1]],
-                        string(l.code),
-                        "Lint.jl",
-                        l.message)
-    end
-    publishDiagnosticsParams = PublishDiagnosticsParams(uri, diags)
-
-    response =  JSONRPC.Request{Val{Symbol("textDocument/publishDiagnostics")},PublishDiagnosticsParams}(Nullable{Union{String,Int64}}(), publishDiagnosticsParams)
+function parse_all(doc, server)
+    # Try blocks should be removed
+    try
+        ps = CSTParser.ParseState(doc._content)
+        doc.code.ast, ps = CSTParser.parse(ps, true)
+    catch er
+        info("PARSING FAILED for $(doc._uri)")
+        info(er)
+    end
+    
+    # includes
+    update_includes(doc, server)
+
+    # diagnostics
+    doc.diagnostics = ps.diagnostics
+
+    # Parsing failed
+    if ps.errored
+        parse_errored(doc, ps)
+    end
+
+    publish_diagnostics(doc, server)
+end
+
+function parse_incremental(doc::Document, dirty::UnitRange, server)
+    isempty(doc.code.ast.args) || sizeof(doc._content) < 800 && return parse_all(doc, server)
+
+    # parsing
+    start_loc = stop_loc = loc = start_block = 0
+    for (i, x) in enumerate(doc.code.ast)
+        if loc < first(dirty) <= loc + x.span || x isa CSTParser.ERROR
+            start_loc = loc
+            start_block = i
+        end
+        if loc < last(dirty) <= loc + x.span
+            stop_loc = loc + x.span
+        end
+        loc += x.span
+    end
+
+    (start_loc == 0 || start_block == 1) && return parse_all(doc, server)
+
+    ps = CSTParser.ParseState(doc._content)
+    # Skip to start position
+    if start_block > 5
+        start_loc1 = sum(doc.code.ast[i].span for i = 1:start_block - 3)
+        skip(ps.l.io, start_loc1)
+        # CSTParser.Tokenize.Lexers.emit(ps.l, CSTParser.Tokenize.Tokens.ERROR)
+    end
+
+    while ps.nt.startbyte < start_loc
+        next(ps)
+    end
+    new_expressions, _ = CSTParser.parse(ps, true)
+
+    # Parsing failed
+    if ps.errored
+        return parse_all(doc, server)
+    end
+    # delete all ast below start point
+    deleteat!(doc.code.ast.args, start_block:length(doc.code.ast))
+    # append new parsing
+    append!(doc.code.ast.args, new_expressions.args)
+    doc.code.ast.span = sizeof(doc._content)
+
+    # get includes
+    update_includes(doc, server)
+
+    # clear diagnostics for re-parsed regions
+    delete_id = []
```

Comment:
If these are type `Any` I think they should be enforced for clarity. And if not then there will be perf boost if you type them out i.e. `Int[]` or similar